---
title: "Teaching Notes for R-Genomics Data Carpentry Workshop"
subtitle: "Software Carpentry Workshop in the Dept. of Human Genetics at the University of Michigan"
author: "Marian L. Schmidt"
date: "October 19 - 20, 2015"
output:
  html_document:
    theme: united
    toc: yes
---

# Before We Start 
Lesson Can be found at <a href="http://tracykteal.github.io/R-genomics/00-before-we-start.html" target="_blank">http://tracykteal.github.io/R-genomics/00-before-we-start.html</a> 

##Learning Objectives  
- Introduce participants to the RStudio interface  
- Set up participants to have a working directory with a data/ folder inside  
- Introduce R syntax  
- Point to relevant information on how to get help, and understand how to ask well formulated questions

##RStudio Interface 
Go over the following:  

- Console  
- Script File
- Environment, History  
- Files, Plots, Packages, Help  

Code and workflow are more reproducible if we can document **EVERYTHING** that we do!  

##Our Goal  
Other people can easily and exactly replicate our workflow and results  

## Getting Started with Our Working Directory  
1. Under `File`, click on `New Project`, choose `New Directory` then `Empty project`.  
2. Enter a name for this new folder called `SWC_HumanGenetics` (known as a **working directory**).  This is where we will work for the rest of the workshop `(~/SWC_HumanGenetics)`.  
3. Click on "Create Project".  
4. Under the `Files` Tab on the right of the screen, click on `New Folder` and create a folder named `data` within `SWC_HumanGenetics` (i.e. `~/SWC_HumanGenetics/data`).  
5. Create a New R Script called `Ecoli_Metagenomes.R` (`File` -> `New File` -> `R Script`) and save it within Your working directory.  
6.  Does everyone's working directory look like this?  

##  Interacting with R  
Two Main ways with Interacting with R:  

1. Using the Console - this is not recommended.  
    + Bottom left panel  
    + Will show commands run and results  
    + Can type commands directly into the console.  **However, they will not be saved when you close the session!**
2.  Using Script Files (Plain text files that contain code).  This is recommended.  
    + Type the code/commands into a script file, send it to the console to check it, and then save the script file.  
  
    
`">"` means that R is ready to receive commands.  You can send code from a script file to the console in 2 ways.  

1. Click the `Run` button in the top right of the script file.  
2. Highlight the desired code and click `Ctrl-Enter`.  
  
A stop sign might show up for longer code.  This shows that R is working to execute the code.  
  
`"+"` means that R is waiting for you to complete the code.  R has not yet run the code.  Usually this is because you forgot to'close' a parenthesis or quotation.  Press `Esc` to get out of trouble.  

## Basics of R  
R is a versatile, open source programming/scripting language that's useful for data analysis/science.  It is:  

- Based on the language S  
- Open Source Software under GPL (a specific type of license).  
- Superior to most alternatives  
- Has over 7,000 contributed packages  
    + Widely used in academia and industry  
- Available on all platforms  
- Not just for statistics, but also for general purpose programming  
- Large and growing community of peers  

## Organizing Your Working Directory  
You should always separate the original, raw data from intermediate datasets you create!  For example your working directory might have folders such as:  

- `raw_data`  
- `intermediate_data`
- `figures`  
- `original_analyses`  
- and so on...  

## Three ways to Seek Help  

1.  You know what function you would like to use but need details about it.  You can type a single question mark:    
```{r help with a function, eval=FALSE, echo=TRUE}
?mean  # Reminding myself of the detail and structure of calling the mean function. 
```
2.  You cannot remember the function that you need to use, but you remember that it has to deal with calculating means.  To search all functions within your R, you can use a double question mark.  This can also help you remind yourself of which package the function is in.    
```{r help search, eval=FALSE, echo=TRUE}
## First way 
??mean  # Searching for a function that has to do with calculting means of some sort. 
## Second way 
help.search("mean")
```
Another way to do this is to 

3. If you need to remind yourself of the names of the arguments:  
```{r arguments, eval=FALSE, echo=TRUE}
args(hist) # "hist" is a function that makes a histogram.
```
Another way is to hit tab.  

### You Get an error message that you don't understand. 

Google it!  Be sure to use key words like **R** and **Error**.  You can copy and paste the error into it.  

### Asking For Help  
- Be sure to make it possible for someone to grasp your problem rapidly.  Make it easy to pinpoint the problem.  
- Try to reproduce the problem with different very simplified data.  You can create a simplified data set by using `dput()`.  

```{r asking for help, eval = TRUE, echo = TRUE}
dput(head(iris)) #Iris is an example data.frame that comes with R.  Head only shows the first 6 rows.  
```
- Always include `sessionInfo()` so people can replicate your example.  

```{r SessionInfo, eval=TRUE, echo=TRUE}
sessionInfo()
```


  
# Introduction to R 
Lesson Can be found at <a href="http://tracykteal.github.io/R-genomics/01-intro-to-R.html" target="_blank">http://tracykteal.github.io/R-genomics/01-intro-to-R.html</a>   

## R Syntax  
*Open up `Final_Analysis.R` and point out the following:*  

- The assignment operator `<-`  
- The `=` for arguments within functions.  
- `#` for comments  
    + This is where you can tell your reader important "behind the scenes" information.  
    + **IMPORTANT:  You should ALWAYS heavily comment your code**
- The `$` operator for working with columns inside of a data frame.  
- Indentation and spacing for easy reading - **Code is made for humans**.  

### A principle:  "Write data for computers and code for humans."  

## Creating objects  
We can do simple math with R.  

```{r simple math uncommented, eval=FALSE, echo=TRUE}
3+5
12/7
```

**What is missing??**  COMMENTS!  ALWAYS comment code 
```{r simple math commented, eval=FALSE, echo=TRUE}
# Adding 3 and 5.  R is fun!
3+5 
``` 
If we forget the `#`, **what happens?**  

- *R is trying to run that sentence as a command*  
- Now R has the `+`  
    + How do we get out of this?  *`Esc`*    
    
To do more interesting things, we now need to learn about **variables and assignment operators.  **

## Objects and Assignment Operators  
For instance instead of adding `3 + 5` we can create objects, which are called **objects**  We assign an object with the `<-` operator, which assigns the value on the right to the object on the left  The `<-` can be read as **"goes into"**    

> **Tip:**  If you use `Alt -` (push `Alt` and then `-`), you can create the `<-` in a single keystroke.  

```{r assign simple variables, eval = FALSE, echo=TRUE}
# assign 3 to a
a <- 3
# assign 5 to b
b <- 5

# what now is a
a
# what now is b
b

#Add a and b 
a + b

#Add a and b and create c
c <- a + b

# what now is c
c
```


> **Tip:** The name of your object is *important*!  It should be meaningful and not too long.  The name of an object can never start with a number. The following are unacceptable names for objects:  

> -`if`, `else`, `for`: These represent fundamental functions within R.  
> -`c`, `mean`, `df`, `data`: These are also functions  
> -You can use `_` and `-` but not `.`  
> -Use nouns for objects and verbs for functions that you write.  
> -Keep a consistent style of your coding.  Remember, code is for humans!  (Hadley Wickham and Google each have published style guides.)  

##Functions and their arguments  
**Functions** are "canned scripts" that automate a task.  They:

- Requires one or more *arguments*, which are the inputs to the function.  
    + Arguments can be anything, numbers, columns of a data.frame, a whole data file.  It differs by the function and can be looked up in the help page.  
    + Many functions have arguments that have *defaults*.  This is a standard value that the author of the function thought would be "good enough in standard cases".  
        + For example, what symbol to use in a plot.  
        + You can always change the default to what you prefer. 
- Often (but not always) return a *value*    
- Executing a function is known as *calling* a function  

For example, let's take the function `sqrt()`:  

- **Argument/Input:** The input, a number.  
- **Output/Return value:** The value input number squared.  

Another example, let's take the function `round()`:  

> **Exercise:**  
> Round Pi:  3.14159   
> 1. How many arguments does round have?  
> 2. What is the default of round?  
> 3. Change the default to a different value.  What happens?

```{r round example, eval=TRUE, echo=TRUE}
# Round pi
round(3.14159) 

# What are the arguments and default of round?
args(round)  

# Changing the default number of digits to 3
round(3.14159, digits = 3)  #Recommended!
round(3.14159, 3) ## NOT recommended for functions with more than one argument!!  
```
We get 3, that's because the default is to round to the nearest whole number.  

> **Exercise:**  
> Create an object named `ecoli_genome_length_mb` with a value of 4.6.  
> How many picograms does the genome weigh? (978 Mbp = 1 pg)  


#####Solution
```{r genome length exercise, eval=TRUE, echo=TRUE}
ecoli_genome_length_mb <- 4.6 # This is for an e-coli genome
ecoli_genome_length_mb
ecoli_genome_weight_pg <- ecoli_genome_length_mb / 978.0
ecoli_genome_weight_pg # Ecoli genomes don't weigh very much. 
```  

#####Human Genome
```{r Human genome, eval=TRUE, echo=TRUE}
human_genome_length_mb <- 30000 # This is for an e-coli genome
human_genome_length_mb
human_genome_weight_pg <- human_genome_length_mb / 978.0 # To calculate the number of picograms that human genomes weight
human_genome_weight_pg # Human genomes don't weigh very much. 
```  

#####Comparing things  
- `==` denotes equality.  Reads as "is equal to"  
- `!=` denotes inequality.  Reads as "is not equal to"  
- `<` "less than"  
- `<=` "less than or equal to"  
- `>` "greater than"  
- `>=` "greater than or equal to"  

```{r comparing things, eval=TRUE, echo=TRUE}
1 ==1 # is 1 equal to 1?
1 != 1 # is 1 inequal to 1?   
1 < 2 # Is 1 less than 2?
```


> **Exercise:**  
> Compare the lengths of `ecoli_genome_length_mb`  and `human_genome_length_mb`. 

```{r compare human and ecoli, eval=TRUE, echo=TRUE}
human_genome_length_mb == ecoli_genome_length_mb # Are the genome lengths equal  

```




