---
title: "Teaching Notes for R-Genomics Data Carpentry Workshop"
subtitle: "Software Carpentry Workshop in the Dept. of Human Genetics at the University of Michigan"
author: "Marian L. Schmidt"
date: "October 19 - 20, 2015"
output:
  html_document:
    theme: united
    toc: yes
---

# Before We Start 
Lesson Can be found at <a href="http://tracykteal.github.io/R-genomics/00-before-we-start.html" target="_blank">http://tracykteal.github.io/R-genomics/00-before-we-start.html</a> 

##Learning Objectives  
- Introduce participants to the RStudio interface  
- Set up participants to have a working directory with a data/ folder inside  
- Introduce R syntax  
- Point to relevant information on how to get help, and understand how to ask well formulated questions

##RStudio Interface 
Go over the following:  

- Console  
- Script File
- Environment, History  
- Files, Plots, Packages, Help  

Code and workflow are more reproducible if we can document **EVERYTHING** that we do!  

##Our Goal  
Other people can easily and exactly replicate our workflow and results  

## Getting Started with Our Working Directory  
1. Under `File`, click on `New Project`, choose `New Directory` then `Empty project`.  
2. Enter a name for this new folder called `SWC_HumanGenetics` (known as a **working directory**).  This is where we will work for the rest of the workshop `(~/SWC_HumanGenetics)`.  
3. Click on "Create Project".  
4. Under the `Files` Tab on the right of the screen, click on `New Folder` and create a folder named `data` within `SWC_HumanGenetics` (i.e. `~/SWC_HumanGenetics/data`).  
5. Create a New R Script called `Ecoli_Metagenomes.R` (`File` -> `New File` -> `R Script`) and save it within Your working directory.  
6.  Does everyone's working directory look like this?  

##  Interacting with R  
Two Main ways with Interacting with R:  

1. Using the Console - this is not recommended.  
    + Bottom left panel  
    + Will show commands run and results  
    + Can type commands directly into the console.  **However, they will not be saved when you close the session!**
2.  Using Script Files (Plain text files that contain code).  This is recommended.  
    + Type the code/commands into a script file, send it to the console to check it, and then save the script file.  
  
    
`">"` means that R is ready to receive commands.  You can send code from a script file to the console in 2 ways.  

1. Click the `Run` button in the top right of the script file.  
2. Highlight the desired code and click `Ctrl-Enter`.  
  
A stop sign might show up for longer code.  This shows that R is working to execute the code.  
  
`"+"` means that R is waiting for you to complete the code.  R has not yet run the code.  Usually this is because you forgot to'close' a parenthesis or quotation.  Press `Esc` to get out of trouble.  

## Basics of R  
R is a versatile, open source programming/scripting language that's useful for data analysis/science.  It is:  

- Based on the language S  
- Open Source Software under GPL (a specific type of license).  
- Superior to most alternatives  
- Has over 7,000 contributed packages  
    + Widely used in academia and industry  
- Available on all platforms  
- Not just for statistics, but also for general purpose programming  
- Large and growing community of peers  

## Organizing Your Working Directory  
You should always separate the original, raw data from intermediate datasets you create!  For example your working directory might have folders such as:  

- `raw_data`  
- `intermediate_data`
- `figures`  
- `original_analyses`  
- and so on...  

## Three ways to Seek Help  

1.  You know what function you would like to use but need details about it.  You can type a single question mark:    
```{r help with a function, eval=FALSE, echo=TRUE}
?mean  # Reminding myself of the detail and structure of calling the mean function. 
```
2.  You cannot remember the function that you need to use, but you remember that it has to deal with calculating means.  To search all functions within your R, you can use a double question mark.  This can also help you remind yourself of which package the function is in.    
```{r help search, eval=FALSE, echo=TRUE}
## First way 
??mean  # Searching for a function that has to do with calculting means of some sort. 
## Second way 
help.search("mean")
```
Another way to do this is to 

3. If you need to remind yourself of the names of the arguments:  
```{r arguments, eval=FALSE, echo=TRUE}
args(hist) # "hist" is a function that makes a histogram.
```
Another way is to hit tab.  

### You Get an error message that you don't understand. 

Google it!  Be sure to use key words like **R** and **Error**.  You can copy and paste the error into it.  

### Asking For Help  
- Be sure to make it possible for someone to grasp your problem rapidly.  Make it easy to pinpoint the problem.  
- Try to reproduce the problem with different very simplified data.  You can create a simplified data set by using `dput()`.  

```{r asking for help, eval = TRUE, echo = TRUE}
dput(head(iris)) #Iris is an example data.frame that comes with R.  Head only shows the first 6 rows.  
```
- Always include `sessionInfo()` so people can replicate your example.  

```{r SessionInfo, eval=TRUE, echo=TRUE}
sessionInfo()
```


  
# Introduction to R 
Lesson can be found at <a href="http://tracykteal.github.io/R-genomics/01-intro-to-R.html" target="_blank">http://tracykteal.github.io/R-genomics/01-intro-to-R.html</a>   

##Learning Objectives  

- Be familiar with R syntax  
- Understand the concepts of objects and assignment  
- Understand the concepts of vector and data types  
- Get exposed to a few functions


## R Syntax  
*Open up `Final_Analysis.R` and point out the following:*  

- The assignment operator `<-`  
- The `=` for arguments within functions.  
- `#` for comments  
    + This is where you can tell your reader important "behind the scenes" information.  
    + **IMPORTANT:  You should ALWAYS heavily comment your code**
- The `$` operator for working with columns inside of a data frame.  
- Indentation and spacing for easy reading - **Code is made for humans**.  

### A principle:  "Write data for computers and code for humans."  

## Creating objects  
We can do simple math with R.  

```{r simple math uncommented, eval=FALSE, echo=TRUE}
3+5
12/7
```

**What is missing??**  COMMENTS!  ALWAYS comment code 
```{r simple math commented, eval=FALSE, echo=TRUE}
# Adding 3 and 5.  R is fun!
3+5 
``` 
If we forget the `#`, **what happens?**  

- *R is trying to run that sentence as a command*  
- Now R has the `+`  
    + How do we get out of this?  *`Esc`*    
    
To do more interesting things, we now need to learn about **variables and assignment operators.  **

## Objects and Assignment Operators  
For instance instead of adding `3 + 5` we can create objects, which are called **objects**  We assign an object with the `<-` operator, which assigns the value on the right to the object on the left  The `<-` can be read as **"goes into"**    

> **Tip:**  If you use `Alt -` (push `Alt` and then `-`), you can create the `<-` in a single keystroke.  

```{r assign simple variables, eval = FALSE, echo=TRUE}
# assign 3 to a
a <- 3
# assign 5 to b
b <- 5

# what now is a
a
# what now is b
b

#Add a and b 
a + b

#Add a and b and create c
c <- a + b

# what now is c
c
```


> **Tip:** The name of your object is *important*!  It should be meaningful and not too long.  The name of an object can never start with a number. The following are unacceptable names for objects:  

> -`if`, `else`, `for`: These represent fundamental functions within R.  
> -`c`, `mean`, `df`, `data`: These are also functions  
> -You can use `_` and `-` but not `.`  
> -Use nouns for objects and verbs for functions that you write.  
> -Keep a consistent style of your coding.  Remember, code is for humans!  (Hadley Wickham and Google each have published style guides.)  

##Functions and their arguments  
**Functions** are "canned scripts" that automate a task.  They:

- Requires one or more *arguments*, which are the inputs to the function.  
    + Arguments can be anything, numbers, columns of a data.frame, a whole data file.  It differs by the function and can be looked up in the help page.  
    + Many functions have arguments that have *defaults*.  This is a standard value that the author of the function thought would be "good enough in standard cases".  
        + For example, what symbol to use in a plot.  
        + You can always change the default to what you prefer. 
- Often (but not always) return a *value*    
- Executing a function is known as *calling* a function  

For example, let's take the function `sqrt()`:  

- **Argument/Input:** The input, a number.  
- **Output/Return value:** The value input number squared.  

Another example, let's take the function `round()`:  

> **Exercise:**  
> Round Pi:  3.14159   
> 1. How many arguments does round have?  
> 2. What is the default of round?  
> 3. Change the default to a different value.  What happens?

```{r round example, eval=FALSE, echo=TRUE}
# Round pi
round(3.14159) 

# What are the arguments and default of round?
args(round)  

# Changing the default number of digits to 3
round(3.14159, digits = 3)  #Recommended!
round(3.14159, 3) ## NOT recommended for functions with more than one argument!!  
```
We get 3, that's because the default is to round to the nearest whole number.  

> **Exercise:**  
> Create an object named `ecoli_genome_length_mb` with a value of 4.6.  
> How many picograms does the genome weigh? (978 Mbp = 1 pg)  


#####Solution
```{r genome length exercise, eval=FALSE, echo=TRUE}
ecoli_genome_length_mb <- 4.6 # This is for an e-coli genome
ecoli_genome_length_mb
ecoli_genome_weight_pg <- ecoli_genome_length_mb / 978.0
ecoli_genome_weight_pg # Ecoli genomes don't weigh very much. 
```  

#####Human Genome
```{r Human genome, eval=FALSE, echo=TRUE}
human_genome_length_mb <- 30000 # This is for an e-coli genome
human_genome_length_mb
human_genome_weight_pg <- human_genome_length_mb / 978.0 # To calculate the number of picograms that human genomes weight
human_genome_weight_pg # Human genomes don't weigh very much. 
```  

#####Comparing things  
- `==` denotes equality.  Reads as "is equal to"  
- `!=` denotes inequality.  Reads as "is not equal to"  
- `<` "less than"  
- `<=` "less than or equal to"  
- `>` "greater than"  
- `>=` "greater than or equal to"  

```{r comparing things, eval=FALSE, echo=TRUE}
1 ==1 # is 1 equal to 1?
1 != 1 # is 1 inequal to 1?   
1 < 2 # Is 1 less than 2?
```


> **Exercise:**  
> Compare the lengths of `ecoli_genome_length_mb`  and `human_genome_length_mb`. 

```{r compare human and ecoli, eval=FALSE, echo=TRUE}
human_genome_length_mb == ecoli_genome_length_mb # Are the genome lengths equal?
human_genome_length_mb != ecoli_genome_length_mb # Are the genome lengths equal?
human_genome_length_mb > ecoli_genome_length_mb # Are human genome lengths greater than ecoli genome lengths?
```


## Vectors and Data Types 
A **vector** is the most common and basic data structure in R.  It's R's workhorse data type.  A **vector** is a list of values, mainly numbers or characters.  You can:  

- Do math with vectors  
- Can assign a list of values to a variable  

Let's create a vector of genome lengths:  

```{r genome lengths vector, echo=TRUE, eval=FALSE}
glengths <- c(4.6, 3000, 50000) # Create a vector of 3 genome lengths
glengths
glengths[1] # Return the first value 
glengths[3] # Return the 3rd value 
```

A vector can also contain characters:
```{r species character vector, echo=TRUE, eval=FALSE}
species <- c("ecoli", "human", "corn")
species
species[2] #What's the 2nd value in the species vector?
```

There are many functions to help us explore the vector: 
```{r vector exploration, echo=TRUE, eval=FALSE}
length(glengths) # How many vlaues are in the genome lengths vector?
length(species) # how many values in the species vector?
length(glengths) == length(species) # Are they the same length? 
```

Some math with the vectors:  
```{r vector math, echo=TRUE, eval=FALSE}
5 * glengths
new_lengths <- 2 * glengths # multiply lengths by 2
new_lengths
```

This makes it easy to combine or work with data!

What type of data is in our vector? 
```{r vector type, echo=TRUE, eval=FALSE}
class(glengths) # What type of data is in glengths?
class(species)
str(glengths) # What's the structure of glengths?
str(species)
```

> **Tip:** `str()` provides an overview of the object and the elements it contains. It is a really useful function when working with large and complex objects:

```{r add values to vector, echo=TRUE, eval=FALSE}
lengths <- c(glengths, 90) # adding at the end
lengths
lengths <- c(30, glengths) # adding at the beginning
lengths
```


We just saw 2 of the 6 data types that R uses: `character` and `numeric`. The other 4 are:

- `logical` for TRUE and FALSE (the boolean data type)  
- `integer` for integer numbers (e.g., 2L, the L indicates to R that it’s an integer)  
- `complex` to represent complex numbers with real and imaginary parts (e.g., 1+4i) and that’s all we’re going to say about them.  
- "raw" that we won’t discuss further.  


Vectors are one of the many data structures that R uses. Other important ones are lists (list), matrices (matrix), data frames (data.frame) and factors (factor).

> Exercise:  
> 1. What's the difference between a *data structure* and *data type*?  
> 2. Discuss the definitions of *vector*, *object*, *function*, *arguments*.

# Starting with Data  
Lesson can be found at <a href="http://tracykteal.github.io/R-genomics/02-starting-with-data.html" target="_blank">http://tracykteal.github.io/R-genomics/02-starting-with-data.html</a> 

##Learning Objectives  

- Load external data (CSV files) in memory using the survey table (Ecoli_metadata.csv) as an example  
- Explore the structure and the content of the data in R  
- Understand what are factors and how to manipulate them

## Looking at Metadata  
We are studying a population of E. coli (designated Ara-3):  

- Propagated for more than 40,000 generations in a glucose-limited minimal medium.  
- This medium was supplemented with citrate which E. coli cannot metabolize in the aerobic conditions of the experiment.  
- Sequencing of the populations at regular time points reveals that spontaneous citrate-using mutants (Cit+) appeared at around **31,000** generations.  
- This metadata describes information on the Ara-3 clones and the columns represent:  

Column  | Description
------------- | -------------
Sample | Clone Name
generation  | generation when sample frozen
clade  | based on parsimony tree
strain  | ancestral strain
cit  | Citrate-using mutant status
run  | Sequence read archive sample ID
genome_size  | size in Mbp (Made up for this lesson)


1.  Make sure you are in the correct working directory by typing `getwd()`.  
2. Create a new directory within this working directory called data.  
3. Move the downloaded file into this directory.  
4. Time to load the data into R!  

```{r load data1, eval=TRUE, echo=TRUE}
getwd() # R's pwd - where are we?!
# reads the Ecoli_metadata.csv file from the data sub-folder
metadata <- read.csv('data/Ecoli_metadata.csv') #provides no output 
```

```{r load data2, eval=FALSE, echo = TRUE}
metadata # This will provide too much output
head(metadata) # first 6 rows
tail(metadata) # last 6 rows 
View(metadata) # View it!
```


## What is a `data.frame`?
`data.frame` is the de facto data structure for most tabular data and what we use for statistics and plotting.  It is:  

- A collection of vectors of identical lengths.  
- Each vector represents a column, and each vector can be of a different data type (e.g., characters, integers, factors).  
- The `str()` function is useful to inspect the data types of the columns.  
- Spreadsheets can be imported with the `read.csv()` or the `read.table` functions.  
- By default, data.frame converts (= coerces) columns that contain characters (i.e., text) into the **factor** data type.  
    + Depending on what you want to do, you might want to keep these columns as characters.  If so, change `read.csv('data/Ecoli_metadata.csv', stringsAsFactors = FALSE)`


## Inspecting `data.frame` objects    

- Size:  
    + `dim()` - returns a vector with the number of rows in the first element, and the number of columns as the second element (the __dim__ensions of the object)  
    + `nrow()` - returns the number of rows  
    + `ncol()` - returns the number of columns  
- Content:  
    + `head()` - shows the first 6 rows  
    + `tail()` - shows the last 6 rows  
    + `View()` - view the data within RStudio
- Names:  
    + `names()` or `colnames()` - returns the column names 
    + `rownames()` - returns the row names  
- Summary:  
    + `str()` - structure of the object and information about the class, length and content of each column  
    + `summary()` - summary statistics for each column

> **Exercises**  
> What is the class of the object metadata?  
> How many rows and how many columns are in this object?  
> How many citrate+ mutants have been recorded in this population?  


##Factors 
**Factors** are used to represent categorical data.   

- Can be ordered or unordered.  
- Are an important class for statistical analysis and for plotting.  
- Are stored as integers, and have labels associated with these unique integers.  
- While factors look (and often behave) like character vectors, they are actually integers under the hood, and you need to be careful when treating them like strings.  
- Once created, factors can only contain a pre-defined set values, known as **levels**.   
    + By default, R always sorts levels in alphabetical order

Let’s now check the __str__ucture of our data.frame in more details with the function `str()`:

```{r structure of metadata, eval=TRUE, echo=TRUE}
str(metadata)
```

- Some columns say `Factor w/ 30 levels`, which means that each row has a unique value.   
- `cit` is a `Factor w/ 3 levels`: `minus`, `plus` and `unknown`.

> **Exercise**  
> By looking at the str(metadata) output:  
> 1. What does the `$` operator mean?  
> 2. What do the integers mean at the end of the description for the columns that are `factors`?

